### PR TITLE
Update reconstructing reducer, version 2.0.0

### PR DESCRIPTION
**Pull request by Protryon**
Uses shift-java like reducer, updates `ConstantFolder` and `BlockSquasher`.

### CODE REVIEW

File: src/main/java/com/shapesecurity/shift/es2017/semantics/Explicator.java [hunk 1]
```diff
@@ -1634,7 +1634,9 @@ private NodeWithValue wrapForWith(String propertyName, Reference reference, Func
 			? _temp_b.x
 			: ('x' in temp_a)
 			  ? _temp_a.x
-			  : x;
+			  :
+
+			  .0;
```
Comment (bakkot) [inline src/main/java/com/shapesecurity/shift/es2017/semantics/Explicator.java]: wat

File: src/main/java/com/shapesecurity/shift/es2017/semantics/visitor/EqualityChecker.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+			return false;
+		}
+		Iterator<Node> node1Iterator = nodeList1.iterator();
+		Iterator<Node> node2Iterator = nodeList2.iterator();
+		while (node1Iterator.hasNext() && node2Iterator.hasNext()) {
+			if (!node1Iterator.next().partiallyEquals(node2Iterator.next())) {
+				return false;
+			}
+		}
+		return !node1Iterator.hasNext() && !node2Iterator.hasNext();
```
Comment (bakkot) [inline src/main/java/com/shapesecurity/shift/es2017/semantics/visitor/EqualityChecker.java:34]: This shouldn't be necessary in light of the length check above, surely?

File: src/main/java/com/shapesecurity/shift/es2017/semantics/visitor/EqualityChecker.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+		}
+		ImmutableList<Node> nodeList1 = Visitor.getChildren(node1);
+		ImmutableList<Node> nodeList2 = Visitor.getChildren(node2);
+		if (nodeList1.length != nodeList2.length) {
+			return false;
+		}
+		Iterator<Node> node1Iterator = nodeList1.iterator();
+		Iterator<Node> node2Iterator = nodeList2.iterator();
+		while (node1Iterator.hasNext() && node2Iterator.hasNext()) {
+			if (!node1Iterator.next().partiallyEquals(node2Iterator.next())) {
```
Comment (bakkot) [inline src/main/java/com/shapesecurity/shift/es2017/semantics/visitor/EqualityChecker.java]: This isn't sufficient, surely? `x = 1 + 2 + 3` should not be equal to `x = 1 + 2 + 4`.

File: src/main/java/com/shapesecurity/shift/es2017/semantics/visitor/EqualityChecker.java (hunk 2 above)
Comment (Protryon) [reply src/main/java/com/shapesecurity/shift/es2017/semantics/visitor/EqualityChecker.java:34]: It's not, but it's cleaner IMO. Helps catch incorrect iterator edge cases I suppose, no reason not to keep it AFAIK.

File: src/main/java/com/shapesecurity/shift/es2017/semantics/visitor/Visitor.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package com.shapesecurity.shift.es2017.semantics.visitor;
+
+import com.shapesecurity.functional.data.ImmutableList;
+import com.shapesecurity.functional.data.Monoid;
+import com.shapesecurity.shift.es2017.semantics.asg.Node;
+
+import javax.annotation.Nonnull;
+
+public final class Visitor extends MonoidalReducer<ImmutableList<Node>> {
```
Comment (bakkot) [inline src/main/java/com/shapesecurity/shift/es2017/semantics/visitor/Visitor.java]: You almost certainly want a `ConcatList` here.

File: src/main/java/com/shapesecurity/shift/es2017/semantics/visitor/Visitor.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+	public static ImmutableList<Node> getChildren(@Nonnull Node node) {
```
Comment (bakkot) [inline src/main/java/com/shapesecurity/shift/es2017/semantics/visitor/Visitor.java]: Can we call this `getDescendants` or something? I usually assume that "children" is just direct children.

**Review by bakkot — COMMENTED** [review]
Works for me.

**Review by bakkot — APPROVED** [review]

**Pull request merged**